test/models/thesis_test.rb

```diff
@@ -395,4 +396,77 @@ class ThesisTest < ActiveSupport::TestCase
     assert t1.holds.count == 0
     assert_difference("Thesis.count", -1) { t1.destroy }
   end
+
+  test 'finds existing thesis from csv' do
+    filepath = 'test/fixtures/files/registrar_data_thesis_existing.csv'
+    row = CSV.readlines(open(filepath), headers: true).first
+    user = users(:yo)
+    user.theses= [theses(:one)]
+    thesis = Thesis.create_or_update_from_csv(user, degrees(:one), departments(:one), Date.new(2017, 9, 1), row)
+    user.reload
+    assert_equal 1, user.theses.size
+    assert_equal theses(:one), thesis
+  end
+
+  test 'creates thesis from csv with expected attributes' do
+    filepath = 'test/fixtures/files/registrar_data_thesis_new.csv'
+    row = CSV.readlines(open(filepath), headers: true).first
+    user = users(:yo)
+    user.theses = []
+    thesis = Thesis.create_or_update_from_csv(user, degrees(:one), departments(:one), Date.new(2017, 9, 1), row)
+    user.reload
+    assert_equal 'Coauthor, Mine', thesis.coauthors
+    assert_equal degrees(:one), thesis.degrees.first
+    assert_equal departments(:one), thesis.departments.first
+    assert_equal 'September', thesis.graduation_month
+    assert_equal 2017, thesis.graduation_year
+    assert_equal 'A New Thesis', thesis.title
+    assert_equal user, thesis.users.first
+    assert_equal 1, user.theses.size
+  end
+
+  test 'updates all expected attributes of existing thesis from csv' do
+    filepath = 'test/fixtures/files/registrar_data_thesis_existing.csv'
+    row = CSV.readlines(open(filepath), headers: true).first
+    thesis = theses(:one)
+    thesis.update(title: '', coauthors: '')
+    user = users(:yo)
+    user.theses = [thesis]
+    degree_count = thesis.degrees.size
+    department_count = thesis.departments.size
+    new_degree = degrees(:two)
+    new_department = departments(:two)
+    Thesis.create_or_update_from_csv(user, new_degree, new_department, Date.new(2017, 9, 1), row)
+    thesis.reload
+    assert_equal 'My new co-author', thesis.coauthors
+    assert_includes thesis.degrees, new_degree
+    assert_equal degree_count + 1, thesis.degrees.size
+    assert_includes thesis.departments, new_department
+    assert_equal department_count + 1, thesis.departments.size
+    assert_equal 'A New Title', thesis.title
+  end
+
+  test 'only updates existing attributes from CSV if needed' do
+    filepath = 'test/fixtures/files/registrar_data_thesis_existing.csv'
+    row = CSV.readlines(open(filepath), headers: true).first
+    thesis = theses(:one)
+    thesis.degrees = [degrees(:two)]
+    thesis.departments = [departments(:two)]
+    user = users(:yo)
+    user.theses = [thesis]
+    Thesis.create_or_update_from_csv(user, degrees(:two), departments(:two), Date.new(2017, 9, 1), row)
+    thesis.reload
+    assert_equal 'My co-author; My new co-author', thesis.coauthors
+    assert_equal thesis.degrees, [degrees(:two)]
+    assert_equal thesis.departments, [departments(:two)]
+    assert_equal 'MyString', thesis.title
+  end
+
```

Comment:
Would you consider adding an explicit test confirming we don't overwrite the title? I think that is implied in this test but that data is so awful in the registrar file I feel it may be worth a distinct test.

Reply:
Since it is included in this test, what if I do explicit assertions for all the things I expect to not change?